### DISCUSSION TITLE
replicating index.add_all with gix

### DISCUSSION THREAD

**Discussion by Lunarequest** [GitoxideLabs/gitoxide#1031] [Q&A] [answered]
to do say `git add .` you would do something like this

```
pub fn git_add(repo: &Repository) -> Result<()> {
    let mut index = repo.index().context("Failed to get index of repo")?;

    index
        .add_all(["."].into_iter(), IndexAddOption::DEFAULT, None)
        .context("Failed to add files to repo")?;
    index.write()?;
    Ok(())
}
```
but I'm a bit confused on how to do the same with gix due to a lack of examples. am i missing something obvious?

**Comment by Byron** [comment, marked as answer]
Currently there is no high-level API to implement `git add` similarly to how `git add` would, or how one would do it [with `git2`](https://docs.rs/git2/0.18.1/git2/struct.Index.html#method.add_all).

As a general rule of thumb, `gix` (the crate) is likely to be able to do what you are looking for if `gix` (the CLI) can do it. Using the CLI for discovery might have advantages as it follows the terminology of the `git` CLI more closely.

Please also note that technically, you can implement a `git add` with `gix` today, but it will be very low-level and cumbersome. Further, it's not yet possible to easily create a tree from an index, which would be needed for a commit.

With this said, it's ~~not yet on the roadmap either as `cargo` doesn't need that functionality~~ on the roadmap as [`cargo`](https://github.com/Byron/gitoxide/issues/106) also needs that to work, but it's unclear if it will be implemented this year.

**Reply by Lunarequest** [reply]
would you mind pointing me to the right file which defines the cli args?

**Reply by Byron** [reply]
What I tried to say is that something like `index.add_all()` does not yet exist, and might not until next year.
